openlibrary/core/bookshelves.py

```diff
@@ -194,26 +241,105 @@ def get_users_logged_books(
             'offset': limit * (page - 1),
             'bookshelf_id': bookshelf_id,
         }
-        if sort == 'created desc':
+
+        # Filtering won't occur unless q >= 3, as limited in mybooks.my_books_view().
+        if q:
+            # Filtering by query needs a larger limit as we need (ideally) all of a
+            # user's added works from the reading log DB. The logged work IDs are used
+            # to query Solr, which searches for matches related to those work IDs.
+            data["limit"] = 30_000
+
             query = (
-                "SELECT * from bookshelves_books WHERE "
+                "SELECT work_id from bookshelves_books WHERE "
                 "bookshelf_id=$bookshelf_id AND username=$username "
-                "ORDER BY created DESC "
-                "LIMIT $limit OFFSET $offset"
+                "LIMIT $limit"
+            )
+
+            reading_log_books = list(oldb.query(query, vars=data))
```

Comment:
An assert might be useful here

```suggestion
            reading_log_books = list(oldb.query(query, vars=data))
            assert len(reading_log_books) <= 30_000
```